### ISSUE TITLE
Nullable false positives when boolean conditional is first assigned to a named variable

### ISSUE THREAD

**Issue by j2jensen** [dotnet/roslyn#52717]
**Version Used**: Microsoft.Net.Compilers.Toolset v3.9.0, or VS/MSBuild 16.9.4

**Steps to Reproduce**:

1. Write code like this. Note the lack of compiler warnings.
```cs
#nullable enable

if(!(new[] { "" }.ToDictionary(i => i.ToString())
    .TryGetValue("", out var value) && value.Any()))
{
    Console.WriteLine("No value");
}
else
{
    Console.WriteLine(value.ToString());
}
```
2. Now extract a named variable:
```cs
#nullable enable

var hasNonEmptyValue = new[] { "" }.ToDictionary(i => i.ToString())
    .TryGetValue("", out var value) && value.Any();
if(!hasNonEmptyValue)
{
    Console.WriteLine("No value");
}
else
{
    Console.WriteLine(value.ToString());
}
```

**Expected Behavior**:
The compiler should know that `value` cannot be null in the `else` block, just as it knows that when not using a variable.

**Actual Behavior**:
CS8602 warning upon calling value.ToString()
Similar behavior with other nullable-related errors if I try to pass the value as an argument, assign it to a field, etc.

**Comment by j2jensen** [comment]
I just realized this is described in [this SO post](https://stackoverflow.com/a/59328672/120955), where Mads Torgensen mentions it might be revisited in C# 9. I'm guessing it wasn't?

**Comment by Youssef1313** [comment]
The current behavior is by design. This may need a detailed proposal on how things should be improved. https://github.com/dotnet/roslyn/issues/27011

**Comment by RikkiGibson** [comment]
It would be great if someone decided to explore this issue in https://github.com/dotnet/csharplang/discussions. In particular, how associating conditional state with boolean variables should work, and how "invalidating" such state should work when changes to the underlying variables occur, i.e.

```cs
void M(string? x)
{
    bool b = x != null;
    x = null; // perhaps conditional state associated with 'b' is now "out of date"
    if (b)
    {
        x.ToString(); // warning
    }
}
```
